File: internal/model/forecast.go
package model

import "time"

// -----------------------------------------------------------------------
// Phase 7: Forecast / AI-predicted RED metrics
// -----------------------------------------------------------------------

// ForecastPoint holds a single timestep of the forecast time-series.
// Actual is nil for future points (predictions only).
type ForecastPoint struct {
	Ts        time.Time `json:"ts"`
	Actual    *float64  `json:"actual"`    // nil for future timestamps
	Predicted float64   `json:"predicted"` // model forecast
	Lower     float64   `json:"lower"`     // confidence interval lower bound
	Upper     float64   `json:"upper"`     // confidence interval upper bound
}

// ForecastMetric is the metric name used in forecast series.
type ForecastMetric string

const (
	ForecastMetricRate      ForecastMetric = "rate"
	ForecastMetricErrorRate ForecastMetric = "error_rate"
	ForecastMetricP95Ms     ForecastMetric = "p95_ms"
)

// ForecastServiceRED carries the forecast time-series for one service/metric pair.
type ForecastServiceRED struct {
	Service string         `json:"service"`
	Metric  ForecastMetric `json:"metric"`
	Unit    string         `json:"unit"`   // "req/min" | "ratio" | "ms"
	Series  []ForecastPoint `json:"series"`
}

// ForecastRedResponse is the envelope for GET /api/v1/forecast/red.
type ForecastRedResponse struct {
	Services     []ForecastServiceRED `json:"services"`
	GeneratedAt  time.Time            `json:"generated_at"`
	HorizonHours int                  `json:"horizon_hours"`
}

// -----------------------------------------------------------------------
// Capacity headroom
// -----------------------------------------------------------------------

// CapacityMetric carries the capacity prediction for one resource/service pair.
type CapacityMetric struct {
	Resource      string    `json:"resource"`       // "cpu" | "memory" | "request_rate"
	Service       string    `json:"service"`
	Current       float64   `json:"current"`        // current utilization
	PredictedMax  float64   `json:"predicted_max"`  // peak within horizon
	Capacity      float64   `json:"capacity"`       // max supported (100 for pct)
	HeadroomPct   float64   `json:"headroom_pct"`   // remaining headroom %
	SaturationAt  *string   `json:"saturation_at"`  // ISO ts when headroom → 0, nil = safe
}

// ForecastCapacityResponse is the envelope for GET /api/v1/forecast/capacity.
type ForecastCapacityResponse struct {
	Metrics      []CapacityMetric `json:"metrics"`
	GeneratedAt  time.Time        `json:"generated_at"`
	HorizonHours int              `json:"horizon_hours"`
}

// -----------------------------------------------------------------------
// Anomaly predictions
// -----------------------------------------------------------------------

// AnomalySeverity is the forecast anomaly severity level.
type AnomalySeverity string

const (
	AnomalySeverityWarn     AnomalySeverity = "warn"
	AnomalySeverityCritical AnomalySeverity = "critical"
)

// AnomalyPrediction holds a single predicted anomaly event.
type AnomalyPrediction struct {
	ID             string          `json:"id"`
	Service        string          `json:"service"`
	Metric         string          `json:"metric"`
	Severity       AnomalySeverity `json:"severity"`
	Description    string          `json:"description"`
	PredictedAt    time.Time       `json:"predicted_at"`    // when anomaly expected to start
	Confidence     float64         `json:"confidence"`      // 0.0–1.0
	CurrentValue   float64         `json:"current_value"`
	ThresholdValue float64         `json:"threshold_value"`
}

// ForecastAnomaliesResponse is the envelope for GET /api/v1/forecast/anomalies.
type ForecastAnomaliesResponse struct {
	Anomalies   []AnomalyPrediction `json:"anomalies"`
	GeneratedAt time.Time           `json:"generated_at"`
}

File: web/src/types/forecast.ts
// Phase 7: Forecast / AI-predicted metrics types

// -----------------------------------------------------------------------
// RED forecast time-series
// -----------------------------------------------------------------------

export interface ForecastPoint {
  ts: string           // ISO timestamp
  actual: number | null  // null for future (predicted-only) points
  predicted: number
  lower: number        // confidence band lower bound
  upper: number        // confidence band upper bound
}

export type ForecastMetric = 'rate' | 'error_rate' | 'p95_ms'

export interface ForecastServiceRED {
  service: string
  metric: ForecastMetric
  unit: string         // "req/min" | "ratio" | "ms"
  series: ForecastPoint[]
}

export interface ForecastRedResponse {
  services: ForecastServiceRED[]
  generated_at: string
  horizon_hours: number
}

// -----------------------------------------------------------------------
// Capacity headroom
// -----------------------------------------------------------------------

export interface CapacityMetric {
  resource: string       // "cpu" | "memory" | "request_rate"
  service: string
  current: number
  predicted_max: number  // peak utilization predicted within horizon
  capacity: number       // maximum supported value (100 for pct)
  headroom_pct: number   // (capacity - predicted_max) / capacity * 100
  saturation_at: string | null  // ISO ts when headroom → 0, null = safe
}

export interface ForecastCapacityResponse {
  metrics: CapacityMetric[]
  generated_at: string
  horizon_hours: number
}

// -----------------------------------------------------------------------
// Anomaly predictions
// -----------------------------------------------------------------------

export type AnomalySeverity = 'warn' | 'critical'

export interface AnomalyPrediction {
  id: string
  service: string
  metric: string
  severity: AnomalySeverity
  description: string
  predicted_at: string   // ISO ts — when anomaly expected to start
  confidence: number     // 0.0–1.0
  current_value: number
  threshold_value: number
}

export interface ForecastAnomaliesResponse {
  anomalies: AnomalyPrediction[]
  generated_at: string
}
